### PR TITLE
- APEX-129 #resolve #comment Fixed bug where tuples can be emitted ou…

### PR DESCRIPTION
**Pull request by ilooner**
…tside of a streaming window

### CODE REVIEW

**Comment by 243826** [comment]
can you add a unit test?

**Comment by ilooner** [comment]
will add a unit test

File: engine/src/main/java/com/datatorrent/stram/engine/InputNode.java [hunk 1]
```diff
@@ -69,24 +69,27 @@ public final void run()
...
     boolean doCheckpoint = false;
+    boolean insideStreamingWindow = false;
 
     try {
       while (alive) {
         Tuple t = controlTuples.sweep();
         if (t == null) {
-          if (insideWindow) {
+          if (insideApplicationWindow) {
             int generatedTuples = 0;
```
Comment (243826) [inline engine/src/main/java/com/datatorrent/stram/engine/InputNode.java:81]: this and line 95 - can move inside of the streaming window block, right?

File: engine/src/main/java/com/datatorrent/stram/engine/InputNode.java [hunk 2]
```diff
@@ -69,24 +69,27 @@ public final void run()
...
+    boolean insideApplicationWindow = applicationWindowCount != 0;
     boolean doCheckpoint = false;
+    boolean insideStreamingWindow = false;
 
     try {
       while (alive) {
         Tuple t = controlTuples.sweep();
         if (t == null) {
-          if (insideWindow) {
+          if (insideApplicationWindow) {
```
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/engine/InputNode.java]: This should be insideStreamingWindow as well. For example handleIdleTime should not be called between streaming window boundaries.

**Comment by ilooner** [comment]
@PramodSSImmaneni @243826 I have added unit tests, I verified that the unit tests failed without the fix, and they now pass with the fix. Please review and merge

Thanks

**Comment by pramodin** [comment]
@243826 do you have any other comments? I was planning on doing the final review.

**Comment by 243826** [comment]
No, please go ahead.

File: engine/src/main/java/com/datatorrent/stram/engine/InputNode.java [hunk 3]
```diff
@@ -145,7 +148,7 @@ else if (PROCESSING_MODE == ProcessingMode.EXACTLY_ONCE) {
 
               ContainerStats.OperatorStats stats = new ContainerStats.OperatorStats();
               reportStats(stats, currentWindowId);
-              if(!insideWindow){
+              if(!insideApplicationWindow){
```
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/engine/InputNode.java:151]: Can't wait till end of application window to collect stats

Comment (otterc) [inline engine/src/main/java/com/datatorrent/stram/engine/InputNode.java:151]: Metrics are collected at the end of application window. I think Tim has just changed the name of the variable and nothing else has changed. @PramodSSImmaneni please go through older discussions about why Metrics are collected at the end of application window. Also I have an improvement where it phase of metrics collections can be specified.

Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/engine/InputNode.java:151]: @chandnisingh They were being collected at end of streaming window before.

Comment (otterc) [inline engine/src/main/java/com/datatorrent/stram/engine/InputNode.java:151]: @PramodSSImmaneni before auto metrics they were. but with auto metrics it was changed and that discussion happened and was discussed.

Comment (otterc) [inline engine/src/main/java/com/datatorrent/stram/engine/InputNode.java:151]: @PramodSSImmaneni  it was documented and shared with the team pre-Apex

**Pull request merged**